x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/rescorer/InferenceRescorerBuilder.java

```diff
@@ -10,71 +10,143 @@
 import org.apache.lucene.util.SetOnce;
 import org.elasticsearch.TransportVersion;
 import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.io.stream.StreamInput;
 import org.elasticsearch.common.io.stream.StreamOutput;
 import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.Rewriteable;
 import org.elasticsearch.index.query.SearchExecutionContext;
 import org.elasticsearch.search.rescore.RescorerBuilder;
 import org.elasticsearch.xcontent.ObjectParser;
 import org.elasticsearch.xcontent.ParseField;
 import org.elasticsearch.xcontent.XContentBuilder;
 import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xpack.core.ClientHelper;
+import org.elasticsearch.xpack.core.ml.action.GetTrainedModelsAction;
+import org.elasticsearch.xpack.core.ml.inference.TrainedModelConfig;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.InferenceConfig;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.InferenceConfigUpdate;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.LearnToRankConfig;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.LearnToRankConfigUpdate;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.StrictlyParsedInferenceConfig;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.ltr.LearnToRankFeatureExtractorBuilder;
+import org.elasticsearch.xpack.core.ml.job.messages.Messages;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.core.ml.utils.NamedXContentObjectHelper;
 import org.elasticsearch.xpack.ml.inference.loadingservice.LocalModel;
 import org.elasticsearch.xpack.ml.inference.loadingservice.ModelLoadingService;
 
 import java.io.IOException;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.function.Supplier;
 
 public class InferenceRescorerBuilder extends RescorerBuilder<InferenceRescorerBuilder> {
 
     public static final String NAME = "inference";
     private static final ParseField MODEL = new ParseField("model_id");
+    private static final ParseField INFERENCE_CONFIG = new ParseField("inference_config");
+    private static final ParseField INTERNAL_INFERENCE_CONFIG = new ParseField("_internal_inference_config");
     private static final ObjectParser<Builder, Void> PARSER = new ObjectParser<>(NAME, false, Builder::new);
     static {
         PARSER.declareString(Builder::setModelId, MODEL);
+        PARSER.declareNamedObject(
+            Builder::setInferenceConfigUpdate,
+            (p, c, name) -> p.namedObject(InferenceConfigUpdate.class, name, false),
+            INFERENCE_CONFIG
+        );
+        PARSER.declareNamedObject(
+            Builder::setInferenceConfig,
+            (p, c, name) -> p.namedObject(StrictlyParsedInferenceConfig.class, name, false),
+            INTERNAL_INFERENCE_CONFIG
```

Comment:
For my education, what's the difference between `InferenceConfig` and `InferenceConfigUpdate`?

It looks like we parse a user's input of `inference_config` field into `InferenceConfigUpdate`?  When do we use `_internal_inference_config` field?


Do we need  `InferenceConfigUpdate` to allow updated to a model?

Reply:
Correct, the `*Update` comes from an API call during the model's inference. The `InferenceConfig` is stored with the model.

The reason for the separation is that certain parts of the config may not be updateable. That is why a different structure underneath is useful. Additionally, users may want "default" behavior for the model that then can be updated later in a different context. Like, calculating feature importance. Feature importance calculations could be `0` by default, but the user could update it to `5` when debugging.

Reply:
Thanks, great explanation!